tests/ctrl/services.go

```diff
@@ -0,0 +1,132 @@
+package ctrl
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Inphi/eip4844-interop/shared"
+	"github.com/ethereum/go-ethereum/ethclient"
+	beaconservice "github.com/prysmaticlabs/prysm/v3/proto/eth/service"
+	"google.golang.org/grpc"
+)
+
+type Service interface {
+	Start(ctx context.Context) error
+	Stop(ctx context.Context) error
+	Started() <-chan struct{}
+}
+
+func NewBeaconNode() Service {
+	url := fmt.Sprintf("http://%s/eth/v1/beacon/genesis", shared.BeaconGatewayGRPC)
+	return newDockerService("beacon-node", url)
+}
+
+func NewValidatorNode() Service {
+	return newDockerService("validator-node", shared.ValidatorRPC)
+}
+
+func GetBeaconNodeClient(ctx context.Context) (beaconservice.BeaconChainClient, error) {
+	// TODO: cache conns for reuse
+	conn, err := grpc.DialContext(ctx, shared.BeaconRPC, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to dial beacon grpc", err)
+	}
+	return beaconservice.NewBeaconChainClient(conn), nil
+}
+
+func NewBeaconNodeFollower() Service {
+	url := fmt.Sprintf("http://%s/eth/v1/beacon/genesis", shared.BeaconGatewayFollowerGRPC)
+	return newDockerService("beacon-node-follower", url)
+}
+
+func GetBeaconNodeFollowerClient(ctx context.Context) (beaconservice.BeaconChainClient, error) {
+	// TODO: cache conns for reuse
+	conn, err := grpc.DialContext(ctx, shared.BeaconFollowerRPC, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to dial beacon follower grpc", err)
+	}
+	return beaconservice.NewBeaconChainClient(conn), nil
+}
+
+func NewGethNode() Service {
+	return newDockerService("execution-node", shared.GethRPC)
+}
+
+func NewGethNode2() Service {
+	return newDockerService("execution-node-2", shared.GethRPC)
+}
+
+func GetExecutionClient(ctx context.Context) (*ethclient.Client, error) {
+	client, err := ethclient.DialContext(ctx, shared.GethRPC)
+	if err != nil {
+		return nil, fmt.Errorf("%w: Failed to connect to the Ethereum client", err)
+	}
+	return client, nil
+}
+
+type dockerService struct {
+	started   chan struct{}
+	svcname   string
+	statusURL string
+}
+
+func (s *dockerService) Start(ctx context.Context) error {
+	if err := StartServices(s.svcname); err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", s.statusURL, nil)
+	if err != nil {
+		return err
+	}
+	for {
+		if _, err := http.DefaultClient.Do(req); err == nil {
+			close(s.started)
+			return nil
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
```

Comment:
is it worth sleeping for say 1 second here?

Reply:
yes. Will do.

Reply:
that might have been the issue in CI! I have another workflow running the 1 second patch that's succeeding!